### ISSUE TITLE
Low-footprint mode

### ISSUE THREAD

**Issue by gmaxwell** [bitcoin/secp256k1#189]
The current strategy is optimized for large fast chips with huge cache. Especially sining would be useful on some embedded devices where multiple megabytes of pre-computation is not acceptable. Reasonably low memory can be accomplished just by lowering window usage, but we could also consider a separate compile-time low-footprint option.

**Comment by sipa** [comment]
Ideas:
* Smaller G table for verification
* Use 16-entry table with only the combinations of b0,b1,b2,b3 for G*(b0*1 + b1*2^64 + b2*2^128 + b3*2^192) (suggested by @gmaxwell as well), needing 64 additions and 64 doublings instead of only additions. (As opposed to the 1024-entry table we have now).
* Use more compact storage format for field entries (generally useful).
* Use generic / branch-free versions of code also where variable-time code is allowed. In particular, avoid 5 different implementations for normalize.
* Selectively only enable compilation of signing or verification.

**Comment by gmaxwell** [comment]
Signing table is now static by default, which delivers a significant fraction of the need here for small devices.

**Comment by dbitbox** [comment]
As a side note to PR #337, is there a security risk for using the wNAF algorithm (i.e. `secp256k1_ecmult_const()`) for signing? This would eliminate the need for the large precomputed context tables. The tradeoff is wNAF runs at half the signing speed:

```
Default setup
./bench_sign
ecdsa_sign: min 83.5us / avg 89.6us / max 95.3us

Using wNAF (with blinding)
./bench_sign 
ecdsa_sign: min 174us / avg 178us / max 182us
```

The verification precomputed context tables (~1.375MB in code comments) can also be eliminated by adding the result of two wNAF `secp256k1_ecmult_const()` calls instead of the current wNAF `secp256k1_ecmult()` implemenation, again running at half the speed. This should make verification small enough to run on embedded devices (not tested).

```
Using secp256k1_ecmult() (default)
./bench_verify 
ecdsa_verify: min 128us / avg 132us / max 135us

Using 2x secp256k1_ecmult_const()
./bench_verify 
ecdsa_verify: min 268us / avg 271us / max 275us
```

Overall, I see 3 different ecmult implementations:  `secp256k1_ecmult()`, `secp256k1_ecmult_gen()`, `secp256k1_ecmult_const()`. Each requires its own precomputation table. Using one ecmult implemenation with the smallest table would be useful for embedded systems. It looks like the table size for wNAF `secp256k1_ecmult_const()` is small at 672 bytes (`secp256k1_ge pre_a[ECMULT_TABLE_SIZE(WINDOW_A)]`) with satisfactory speeds.

I am happy to give it a go if it makes sense.

**Comment by gmaxwell** [comment]
Changing to 2x secp256k1_ecmult_const() is a 2.36x slowdown on my laptop-- does your version pass the tests?

ecdsa_verify: min 57.2us / avg 57.4us / max 57.7us
ecdsa_verify: min 136us / avg 136us / max 137us

Probably better would be ecmult_gen + ecmult_const for this-- for me thats taking only 106 microseconds, so only a 1.84x slowdown; and is sharing the same table with signing.

> It looks like the table size for wNAF secp256k1_ecmult_const() is small at 672 bytes

Take care not to conflate ROM tables with RAM ones, ecmult also only uses a window A sized dynamic table, the rest is constant.  For many cases this makes a _big_ difference.

**Comment by dbitbox** [comment]
Thank you for the detailed reply.

> In verify; Changing to 2x secp256k1_ecmult_const() is a 2.36x slowdown on my laptop-- does your version pass the tests?

Yes, it passes the tests. I just checked again and got the same performance numbers. This is the quick hack I used: https://github.com/dbitbox/secp256k1/blob/smallverify/src/ecmult_impl.h#L269

A 32kB ecmult_gen table would be workable for me. Then an ecmult_gen + ecmult_const sounds attractive for verification.

One other situation to give motivation for making the code size as small as possible is the use case of verifying the authenticity of signed firmware updates. In this case, the verification code would need to share space inside a bootloader, which of course should be small in order to give more room for the firmware.

**Comment by vhnatyk** [comment]
Hi - any progress on this? I see it's hanging for soo long. Need it desperately for an embedded device with 256KiB RAM and 1MiB flash and trying to find the right way and balance to reduce RAM footprint of ecmult precompute - as it tries to fill all RAM and kills the app.

**Comment by real-or-random** [comment]
@vhnatyk 
Have you checked the latest revisions? We made a lot of progress for use on embedded devices in the past weeks (PRs #557, #566, #595, #596). In particular you can configure the size of the ecmult table used for verification (#596) and building the table does not require a lot of additional RAM now (PR #557).

**Comment by real-or-random** [comment]
This issue is related to #622.

**Comment by vhnatyk** [comment]
@real-or-random Yes - sorry that was a bit lazy question in hopes someone competent and focused would reply. 

> Have you checked the latest revisions?

Y, by the time of your reply I got some bits of info from connected PRs about WINDOW_A and WINDOW_G tuning 😄 but, since this page was the only I got after a day or two of googling, hopefully, your reply as a contributor will same somebody else's good few hours, not to mention a couple more of research through the confusing tree of PR's.

Btw I came here from [rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1) noticed you there as well. Hopefully, it will get update sources soon as I see PR is open already. Not sure I'm fit for PR or review, but going to work on that and willing to help  - saw there some discussion/PR's started in [rust-secp256k1#115](https://github.com/rust-bitcoin/rust-secp256k1/pull/115) - will continue there. Thanx 👍

**Comment by elichai** [comment]
Isn't it possible to verify/sign without any precumputed tables? 
Just manually doing all the needed multiplications

**Comment by vhnatyk** [comment]
> Isn't it possible to verify/sign without any precumputed tables?
> Just manually doing all the needed multiplications

Yes, probably - but what about safety with constant time and/or performance. Hmm need to find out with tests

**Comment by elichai** [comment]
hmm I'm not big on side channel attacks,
if you do all the needed multiplications it will take a different time depending on the private key. 
can you leverage that? and is that not the same even with tables? like an odd vs even private key should take different time to compute the pubkey.

**Comment by sipa** [comment]
@elichai There really isn't any need for that. We have perfectly fine variable time and constant time algorithms. There is also no need to run without tables entirely; just make them small enough so it's ok (you can very reasonably make them just a few kB or even less). Doing so will still be vastly more efficient than using algorithm with no tables at all.

**Comment by vhnatyk** [comment]
> There is also no need to run without tables entirely; just make them small enough so it's ok (you can very reasonably make them just a few kB or even less). Doing so will still be vastly more efficient than using algorithm with no tables at all.

Yes, exactly 👍  - the problem I couldn't make it work with all the defines like `ECMULT_WINDOW_SIZE` setting to 2 etc. I'll try to find exactly why it fails on mcu, and if will find out something meaningful  will share here, thanx for help 😄

**Comment by jonasnick** [comment]
#1058 was merged and I believe it addresses this issue.
